File: snippets/go/usecases/workflows/workflowactions.go
package workflows

import (
	"time"

	restate "github.com/restatedev/sdk-go"
)

type OrderDetails struct {
	ID     string `json:"id"`
	Amount int    `json:"amount"`
}

type OrderStatus struct {
	Payment *string       `json:"payment"`
	Order   *OrderDetails `json:"order"`
}

type ApprovalWorkflow struct{}

func (ApprovalWorkflow) Run(ctx restate.WorkflowContext, user User) (string, error) {
	order := OrderDetails{ID: "order123", Amount: 1500}

	// <start_state>
	// Store intermediate results
	restate.Set(ctx, "payment-status", "completed")
	restate.Set(ctx, "order-details", order)
	// <end_state>

	// <start_approval>
	// Wait for external approval
	approval, err := restate.Promise[bool](ctx, "manager-approval").Result()
	// <end_approval>
	if err != nil {
		return "nil", err
	}

	_ = approval

	// <start_timers>
	// Wait for user action with timeout
	userResponse := restate.Promise[string](ctx, "user-response")
	timeout := restate.After(ctx, 24*time.Hour)

	selector := restate.Select(ctx, userResponse, timeout)
	switch selector.Select() {
	case userResponse:
		return userResponse.Result()
	case timeout:
		return "timed out", nil
	}
	// <end_timers>

	return "result", nil
}

// <start_approve>
// External system resolves the promise
func (ApprovalWorkflow) Approve(ctx restate.WorkflowSharedContext, decision bool) error {
	return restate.Promise[bool](ctx, "manager-approval").Resolve(decision)
}

// <end_approve>

type OrderWorkflow struct{}

// <start_state_get>
// Query from external handler
func (OrderWorkflow) GetOrderDetails(ctx restate.WorkflowSharedContext) (OrderDetails, error) {
	return restate.Get[OrderDetails](ctx, "order-details")
}

// <end_state_get>

File: snippets/python/src/usecases/workflows/workflow_actions.py
import restate
from datetime import timedelta
from typing import Dict, Any, Optional
from pydantic import BaseModel


class User(BaseModel):
    name: str
    email: str


class OrderDetails(BaseModel):
    id: str
    amount: int


approval_workflow = restate.Workflow("workflow-actions")


@approval_workflow.main()
async def run(ctx: restate.WorkflowContext, user: User) -> Dict[str, bool]:
    order = OrderDetails(id="order123", amount=1500)

    # <start_state>
    # Store intermediate results
    ctx.set("payment-status", "completed")
    ctx.set("order-details", order)
    # <end_state>

    # <start_approval>
    # Wait for external approval
    approval = await ctx.promise("manager-approval")
    # <end_approval>

    # <start_timers>
    # Wait for user action with timeout
    match await restate.select(
        user_response=ctx.promise("user-response").value(),
        timeout=ctx.sleep(timedelta(days=1)),
    ):
        case ["user_response", response]:
            print("Response:", response)
        case _:
            print("Timeout occurred")
    # <end_timers>

    return {"success": True}


# <start_approve>
# External system resolves the promise
@approval_workflow.handler()
async def approve(ctx: restate.WorkflowSharedContext, decision: bool):
    await ctx.promise("manager-approval").resolve(decision)


# <end_approve>

order_workflow = restate.Workflow("order-workflow")


# <start_state_get>
# Query from external handler
@order_workflow.handler()
async def get_order_details(ctx: restate.WorkflowSharedContext) -> OrderDetails:
    return await ctx.get("order-details")


# <end_state_get>
